crates/ide-db/src/imports/import_assets.rs

```diff
@@ -536,6 +537,31 @@ fn trait_applicable_items(
         required_assoc_items.insert(assoc);
         Some(assoc_item_trait.into())
     })
+    .filter(|&candidate_trait_id| {
+        // we care about the following cases:
+        // 1. Trait's definition crate
+        // 2. Definition crates for all trait's generic arguments
+        //     a. This is recursive for fundamental types: `Into<Box<A>> for ()`` is OK, but
+        //        `Into<Vec<A>> for ()`` is *not*.
+        // 3. Receiver type definition crate
+        //    a. This is recursive for fundamental types
```

Comment:
How does this implement the recursive parts? It seems from the tests that it works, yet I'm struggling to see where you check the crates. I only see checks for two crates here, not the recursive cases...

Reply:
> How does this implement the recursive parts?

I'm not entirely sure! I found that the tests just worked with the expected behavior? I assume this occurs due to the implementation of `TyFingerprint` inside of `hir_ty`, but to be entirely honest, I don't fully understand that code yet.

Just to confirm, when you say recursive, are you thinking of a case like:

```rust
impl foreign::TestTrait for fundamental::Box<fundamental::Box<local::TestStruct>> {
    fn some_method(&self) {}
}
```

...where the method receiver is a double-Boxed `TestStruct`?

Reply:
Yes, that's what I mean. Although you don't even need double boxes, just a single `Box<TestStruct>` will make you check 3 crates (foreign, fundamental, local)...

Reply:
I think that's been implemented in the test named `trait_method_fuzzy_completion_aware_of_fundamental_boxes` :)

Reply:
I know, thus 
> It seems from the tests that it works

but I can't see why

Reply:
ah, i misunderstood, sorry! I interpreted " I only see checks for two crates here, not the recursive cases..." as "I don't see tests for the recursive case", not "it's not obvious to me as to why this works".